pxr/usd/usdPhysics/rigidBodyAPI.cpp

```diff
@@ -250,3 +250,402 @@ PXR_NAMESPACE_CLOSE_SCOPE
+        GfMatrix3f inMatr(0.0f);
+        inMatr[0][0] = shapeMassInfo.diagonalInertia[0];
+        inMatr[1][1] = shapeMassInfo.diagonalInertia[1];
+        inMatr[2][2] = shapeMassInfo.diagonalInertia[2];
+
+        inertia = inMatr * rotMatr;
+    }
+
+    if (shapeMassInfo.hasPa)
+    {
+        inertia = MassProperties::RotateInertia(inertia, shapeMassInfo.principalAxes);
+    }
+
+    // center of mass provided, update the inertia
+    if (hasCoM)
+    {
+        if (!shapeMassInfo.hasInertia)
+        {
+            // update inertia if we override the CoM but use the computed inertia
+            MassProperties massProps(shapeMassInfo.mass, inertia, massInfo.centerOfMass);            
+            massProps.Translate(centerOfMass - massProps.GetCenterOfMass());
+            inertia = massProps.GetInertiaTensor();
+        }
+        massInfo.centerOfMass = centerOfMass;
+    }
+
+    // setup collision transformation
+    transform.SetTranslate(massInfo.localPos);
+    transform.SetRotateOnly(GfQuatd(massInfo.localRot));
+
+    // return final collision mass properties
+    return MassProperties(shapeMassInfo.mass, inertia, massInfo.centerOfMass);
+}
+
+// compute mass properties for given rigid body
+float UsdPhysicsRigidBodyAPI::ComputeMassProperties(GfVec3f& _diagonalInertia, GfVec3f& _com, GfQuatf& _principalAxes, const MassInformationFn& massInfoFn) const
+{
+    MassProperties massProps;
+
+    const UsdPrim usdPrim = GetPrim();
+    const UsdStagePtr stage = usdPrim.GetStage();
+
+    UsdGeomXformCache xfCache;
+
+    // Parse dynamic body mass data
+    const MassApiData rigidBodyMassInfo = ParseMassApi(usdPrim);
+
+    // If we dont have mass, we need to compute mass from collisions
+    const bool accumulateMass = rigidBodyMassInfo.mass <= 0.0f;
+
+    // Get initial data from parsed rigid body mass information
+    GfVec3f outCenterOfMass(0.0f);
+    const bool hasPa = rigidBodyMassInfo.hasPa;
+    GfQuatf outPrincipalAxes = rigidBodyMassInfo.principalAxes;
+    const bool hasCoM = GetCoM(usdPrim, outCenterOfMass, xfCache);    
+    float outMass = rigidBodyMassInfo.mass;
+    GfVec3f outDiagonalizedInertiaTensor = rigidBodyMassInfo.diagonalInertia;
+
+    // if we dont have enough mass information, we need to traverse collisions to gather final mass information
+    if (accumulateMass || !rigidBodyMassInfo.hasInertia || !hasCoM)
+    {
+        std::vector<MassProperties> massProps;
+        std::vector<GfMatrix4f> massTransf;
+        std::vector<UsdPrim> collisionPrims;        
+
+        // traverse all collisions below this body and get their collision information
+        // first gather all collisions
+        UsdPrimRange range(usdPrim);
+        for (auto it = range.begin(); it != range.end(); ++it)
+        {
+            const UsdPrim collisionPrim = *it;
+            if (collisionPrim && collisionPrim.HasAPI<UsdPhysicsCollisionAPI>())
+            {
+                collisionPrims.push_back(collisionPrim);
```

Comment:
If you remove the `const` above, which isn't buying you much, then you can `std::move(collisionPrim)` here, which _does_ buy you several refcount increments.